Submitted By: Joe Ciccone <jciccone@gmail.com>
Date: 2011-01-30
Initial Package Version: 3.2.7
Origin: Fedora Git
        http://pkgs.fedoraproject.org/gitweb/?p=procps.git;a=blob_plain;f=procps-3.2.7-ps-cgroup.patch;h=b638c4f9a826040099f10e9282aad0d31ec5f03f;hb=HEAD
        http://pkgs.fedoraproject.org/gitweb/?p=procps.git;a=blob_plain;f=procps-3.2.8-ps-cgroup-suppress-root-group.patch;h=4b192c1fb4aa7684b7e57606713ca68690d77845;hb=HEAD
Upstream Status: Unknown
Description: Adds the ability for ps to output process(es) cgroups.

diff -Naur procps-3.2.8.orig/ps/output.c procps-3.2.8/ps/output.c
--- procps-3.2.8.orig/ps/output.c	2009-05-10 15:24:50.000000000 -0400
+++ procps-3.2.8/ps/output.c	2011-01-30 12:58:30.530497002 -0500
@@ -1099,6 +1099,46 @@
   return snprintf(outbuf, COLWID, "*");
 }
 
+static int pr_cgroup(char *restrict const outbuf, const proc_t *restrict const pp){
+  char filename[48];
+  FILE *fd;
+  int counter = 0, last_begin = 0, slash = 0;
+  int c;
+  int is_cgroup = 0;
+
+  outbuf[0]='\0';
+  snprintf(filename, sizeof filename, "/proc/%d/cgroup", pp->tgid);
+  fd = fopen(filename, "r");
+  if (likely(fd == NULL)) goto fail;
+  while (( (c = fgetc(fd)) != EOF) && (counter<665)) {
+    if (is_cgroup == 0) {
+      if (c == ':') {
+        is_cgroup = 1;
+        slash = 0;
+        if (counter>0)
+          outbuf[counter++]=';';
+      }
+    }else
+      if ((c == '\n') || (c == '\0')){
+        if (slash)  /* if the last char was a / this process is in the root cgroup which we should suppress */
+            counter = last_begin;
+      else
+            last_begin = counter;
+        is_cgroup = 0;
+      } else {
+        slash = c == '/';
+        outbuf[counter++]=c;
+  }
+  }
+  outbuf[counter]='\0';
+  close(fd);
+  if (counter>0) 
+    return counter;
+fail:
+  outbuf[0] = '-';
+  outbuf[1] = '\0';
+  return 1;
+}
 
 /****************** FLASK & seLinux security stuff **********************/
 // move the bulk of this to libproc sometime
@@ -1293,6 +1333,7 @@
 {"bsdtime",   "TIME",    pr_bsdtime,  sr_nop,     6,   0,    LNX, ET|RIGHT},
 {"c",         "C",       pr_c,        sr_pcpu,    2,   0,    SUN, ET|RIGHT},
 {"caught",    "CAUGHT",  pr_sigcatch, sr_nop,     9,   0,    BSD, TO|SIGNAL}, /*sigcatch*/
+{"cgroup",    "CGROUP",  pr_cgroup,   sr_nop,    35,   0,    LNX, PO|LEFT},  /* cgroups*/
 {"class",     "CLS",     pr_class,    sr_sched,   3,   0,    XXX, TO|LEFT},
 {"cls",       "CLS",     pr_class,    sr_sched,   3,   0,    HPU, TO|RIGHT}, /*says HPUX or RT*/
 {"cmaj_flt",  "-",       pr_nop,      sr_cmaj_flt, 1,  0,    LNX, AN|RIGHT},
diff -Naur procps-3.2.8.orig/ps/ps.1 procps-3.2.8/ps/ps.1
--- procps-3.2.8.orig/ps/ps.1	2009-05-10 15:38:17.000000000 -0400
+++ procps-3.2.8/ps/ps.1	2011-01-30 12:58:08.382497001 -0500
@@ -904,6 +904,10 @@
 displayed.  (alias\ \fBsig_catch\fR,\ \fBsigcatch\fR).
 T}
 
+cgroup	CGROUP	T{
+display control groups to which the process belonges.
+T}
+
 class	CLS	T{
 scheduling class of the process.  (alias\ \fBpolicy\fR,\ \fBcls\fR).
 Field's possible values are:
